### PR TITLE
Paalman & Pings Monte Carlo Absorption

### LINKED ISSUE #28353: Monte Carlo Absorption v2
Currently MonteCarloAbsorption algorithm outputs one correction workspace corresponding to the self attenuation of the ensemble of the sample and [optionally] its environment. Better would be to calculate all 4 terms following the Paalman & Pings formalism (https://docs.mantidproject.org/nightly/algorithms/ApplyPaalmanPingsCorrection-v1.html).

This means a v2 of `MonteCarloAbsorption`, and v2 of `CalculateMonteCarloAbsorption`.
`SimpleShapeMonteCarloAbsorption` will be out of use then.

### LINKED ISSUE #28535: NaN's in sample logs
I have been told that if a sample log contains a NaN this can causes problems when loading. I have been told of an error that says

"format of file not supporting NaN's"

We should talk to Freddie about this,

### PR DESCRIPTION
**Pull request by gvardany**
**Description of work.**
Adds a new algorithm to calculate the attenuation corrections in Paalman & Pings formalism using Monte Carlo Absorption. Supports direct and indirect geometry inelastic and elastic modes. Supports shapes flat plate, cylinder and hollow cylinder in respective containers.
Superseeds and deprecates `CalculateMonteCarloAbsorption` and `SimpleShapeMonteCarloAbsorption`.

**To test:**

Replay the examples in the system test.

Fixes #28353  #28341 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by SpencerHowells** [comment]
@gvardany Great news & work! What we have been doing here with major developments is providing me with a build so that I can do scientific tests before it goes to Review & Approval. Can you do that?

**Comment by gvardany** [comment]
Sure, I was planning to do so, it would be indeed nice to test thoroughly all the shapes.

**Review by DannyHindson — CHANGES_REQUESTED** [review]
I've had an initial look through this and attached a few comments. Might come back and have a look through the tests later on.

Are you planning on leaving the CalculateMonteCarloAbsorption  and SimpleShapeMonteCarloAbsorption algorithms in place? Not much experience of what happens to algorithms when they are deprecated but wondered if there was a change needed to the documentation at least to indicate this.

Also there are a couple of other pieces of documentation that need to be added:
- documentation for the new PaalmanPingsMonteCarloAbsorption algorithm
- the documentation for the SetSample algorithm needs some updates for the new container parameters and the new shapes that are supported: FlatPlateHolder\HollowCylinderHolder

**Comment by SpencerHowells** [comment]
@gvardany @DannyHindson regarding the deprecation of the simpler algorithms our users have a habit of wanting to ‘do their own thing’ and write their own routines. So I’d like them kept for the time being

**Comment by gvardany** [comment]
That's right, I still need to write the documentations for SetSample and the new algorithm.
For the 2 others, they won't be removed yet, but I'll make them deprecated, so that at execution they display a warning suggesting to use the new one instead. There is no much point in keeping the simpler ones, since you can do the simple calculation also with the new one, if you do not give the container. As for the `MaterialAlreadyDefined`, it is not working right now, I will think about it. I'll address the rest later on.

File: Framework/DataHandling/src/SetSample.cpp
Comment (gvardany) [reply Framework/DataHandling/src/SetSample.cpp:645]: That's a good point, ideally these properties should say MassDensity and NumberDensity, since it does not matter it's about the sample or anything else. However I did not change this for not breaking the API. One could prefix them with Container, but that will need a tweak in the validation I guess.

File: Framework/DataHandling/src/SetSample.cpp
Comment (gvardany) [reply Framework/DataHandling/src/SetSample.cpp:618]: That's a possibility; I did not change that again for not accidentally breaking the API, since the validation that I am doing here is far more strict. Actually it would be neater to use the child algorithm for both, but I'm not sure if SetSampleMaterial can handle the container.

File: Framework/DataHandling/src/SetSample.cpp
Comment (gvardany) [reply Framework/DataHandling/src/SetSample.cpp:422]: Yes, I can.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PaalmanPingsMonteCarloAbsorption.py
Comment (gvardany) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PaalmanPingsMonteCarloAbsorption.py:355]: I thought about this, yes we should find a workaround. Just to be on the safe side I clear the sample object before every calculation to make sure there are no historic remnants, but this won't work in your case.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PaalmanPingsMonteCarloAbsorption.py
Comment (gvardany) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PaalmanPingsMonteCarloAbsorption.py:325]: I think the best would be to assume by default that both shape and material are specified.
And only if one of them is not, then the user input properties should become mandatory.

File: Framework/DataHandling/src/SetSample.cpp
Comment (DannyHindson) [reply Framework/DataHandling/src/SetSample.cpp:640]: If you kept the logic for "SampleNumberDensity" and added the following lines in setMaterial I don't think there would be any negative impact on existing scripts that call SetSample:
```
  if (materialArgs.existsProperty("ContainerNumberDensity")) { 
    materialParams.sampleNumberDensity = 
        materialArgs.getProperty("ContainerNumberDensity"); 
  }
```
...or you could alternatively support "NumberDensity" if you prefer.

You could also change the field name on the materialParams struct to "numberDensity" but this would require a few changes to other units that use this struct so this is probably lower priority.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PaalmanPingsMonteCarloAbsorption.py
Comment (DannyHindson) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PaalmanPingsMonteCarloAbsorption.py:325]: Do you mean that the algorithm should assume by default that the shape and material are already specified in the sample\container workspaces that the user supplies? This would be fine with me and it would avoid adding an extra parameter to act as the switch I mentioned in my earlier comment

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PaalmanPingsMonteCarloAbsorption.py
Comment (DannyHindson) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PaalmanPingsMonteCarloAbsorption.py:355]: I'm pretty sure that my suggestion above (changes to SetSample.cpp and MCInteractionVolume.cpp) will allow MonteCarloAbsorption to run without a sample shape defined. I guess where I said "If that doesn't work..." I really meant ("If you don't like that..."). 
If you're v busy I can make these changes in a separate PR - I'm aware this is mostly to benefit my work. That would still leave a small change to the PaalmanPings algorithm to remove the workaround where the container shape is put into the sample which either of us could do depending on which PR went through first.

Comment (gvardany) [reply Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PaalmanPingsMonteCarloAbsorption.py:355]: I'll take care of this, it's just that I was exploring the python API of sample, and there does not seem to be a set cross section, but I will find a way.

File: docs/source/algorithms/SetSample-v1.rst
Comment (DannyHindson) [inline docs/source/algorithms/SetSample-v1.rst:35]: Can the "Environment Definition File" and "Container Geometry and Material" headers be made smaller than the "Environment" header? Think you need to use something other than ----- which is already used for the top level Description header

File: docs/source/algorithms/SetSample-v1.rst
Comment (DannyHindson) [inline docs/source/algorithms/SetSample-v1.rst:71]: The documentation for the material\geometry inputs seems to be combined for sample and container. But this sits below a header called "Container Geometry and Material". Maybe you need to have a section for "Geometry" and "Material" and then refer back to these from the "Container Geometry and Material" section? It's quite nice if there's a header for each of the input parameters

File: docs/source/algorithms/SetSample-v1.rst
Comment (DannyHindson) [inline docs/source/algorithms/SetSample-v1.rst:97]: There are a couple of problems with the markup in the FlatPlateHolder and HollowCylinderHolder sections (yellow highlights are mine):

SetSampleDocScreenshot

**Review by DannyHindson — CHANGES_REQUESTED** [review]
Just had a look through the new SetSample documentation and have added 3 comments. Think that's everything from me for now

**Comment by gvardany** [comment]
I thought about the shape and material already defined issue. It turns out it is more complicated than I thought. The issue is that one can have something defined, but still willing to override it.
Moreover the presence of container has to be treated with care, since a predefined container will mess up with the sample only correction and other way around. So I think it's best if we do this in a separate PR and here assume that we are always starting from scratch. This is even more delicate for shapes since there are no sample and container shapes separately, but if the sample is let's say a flat plate, container is assumed to be a flat plate holder, whereas if we let predefined containers it can be anything. Are you happy with this plan? I think it will also be easier to do this extension later, when we will already have tests for this one. On the other hand, I added the A-ss correction also in the case of container workspace given, just to be consistent with the other paalman pings algorithms. I also improved the set sample documentation. I still need to write the documentation for the new algorithm.

**Comment by DannyHindson** [comment]
Yes - I just thought this through a bit more myself and I can see the complexity. If we allow users to load workspaces with shapes\materials already attached it's not v clear whether the can and sample shapes should be attached to the both the sample ws and container ws. The user may assume the container ws should only have the container shape\material etc... Fine to leave this for a separate PR.

I think eventually it would be good to set this up so that a user can use the full functionality in SetSample\LoadSampleShape\LoadSampleEnvironment to specify the various shapes\materials that get used in PaalmanPingsMonteCarloAbsorption - this would allow mesh geometries to be used but it would also allow arbitrary CSG objects to be specified as well.
Perhaps we could eventually add a 3rd input workspace parameter to PaalmanPingsMonteCarloAbsorption that supplies the shapes\materials? This would allow the user to set up the materials\shapes on that workspace first by manually running SetSample\LoadSampleShape\LoadSampleEnvironment before calling PaalmanPingMonteCarloAbsorption. It would approximately follow the current approach where it is assumed that the sample ws and container ws don't have any shapes attached.

We'd also discussed a change to MonteCarloAbsorption to allow it run without a sample shape (this would avoid the need for the trick where the sample shape is populated with the container shape during the acc calc). Are you also thinking of leaving this to a separate PR? Fine if yes - just checking what's left to do on this one.

**Comment by gvardany** [comment]
Yes, that as well I would leave for the other PR. Basically here it clears the sample before configuring it, so any previously setup shapes/geometries are lost. I have opened the issue #28803. Not sure about the 3rd workspace, but embedding the full functionality of SetSample, etc. is not a great idea either. I think still the best would be if this supports predefined shapes/materials, but this needs to be implemented with care. Perhaps, adding explicit switches 
will reduce the confusion to some extent. One might also stretch the API of sample and material to allow for copying them over and setting in python, I have the impression the current API is rather poor.

**Comment by gvardany** [comment]
With the above, I think it's only documentation that is due for this PR. I will also ask Spencer for the scripts of the examples to turn into system tests.

**Review by DannyHindson — CHANGES_REQUESTED** [review]
I had a look through your latest commits. I've added a few minor requests about SetSample and its documentation page but rest looks good. I'll have another look once you've done the documentation\system tests.
One of my users is keen to run some calculations in MonteCarloAbsorption using the "sample only" option that you've added in so if we could get this merged this week that would be great.

**Comment by gvardany** [comment]
I think that's all now. I will build a package from the windows build, so that Spencer can have a quick test again, and we should be able to push it.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PaalmanPingsMonteCarloAbsorption.py
Comment (DannyHindson) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/PaalmanPingsMonteCarloAbsorption.py:75]: Perhaps would be helpful to expand the doc string to explain this workspace contains measurement with sample and container present.

**Review by DannyHindson — CHANGES_REQUESTED** [review]
Looks like there are a couple of build problems still but I had a look at the new documentation on the main algorithm anyway. I've added two comments. I don't have anything other comments so happy to approve when these are complete. 
Note - I have another PR (#28795) which addresses a problem with the logging that changes the interface to MCAbsorptionStrategy (again). I'm hoping yours goes through first so you don't have to address a merge conflict.

**Review by DannyHindson — APPROVED** [review]
Looks great now thanks

**Comment by martyngigg** [comment]
@SpencerHowells As this was going to conflict with other work we've merged it so it'll be in the next nightly for testing. We can open other issues to address anything that needs looking into.

**Pull request merged**
